src/V4Router.sol

```diff
@@ -0,0 +1,165 @@
+                Currency currency;
+                address recipient;
+                assembly ("memory-safe") {
+                    currency := calldataload(params.offset)
+                    recipient := calldataload(add(params.offset, 0x20))
+                }
+
+                int256 delta = poolManager.currencyDelta(address(this), currency);
+                if (delta < 0) revert InvalidDeltaForAction();
+
+                // TODO should _take have a minAmountOut added slippage check?
+                // TODO recipient mapping
+                _take(currency, recipient, uint256(delta));
+            } else {
+                revert UnsupportedAction(action);
+            }
+        }
+    }
+
+    function _swapExactInputSingle(IV4Router.ExactInputSingleParams memory params) private {
+        _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(-int128(params.amountIn)),
+            params.sqrtPriceLimitX96,
+            params.hookData
+        );
+    }
+
+    function _swapExactInput(IV4Router.ExactInputParams memory params) private {
+        unchecked {
+            // Caching for gas savings
+            uint256 pathLength = params.path.length;
+            uint128 amountOut;
+            uint128 amountIn = params.amountIn;
+            Currency currencyIn = params.currencyIn;
+            PathKey memory pathKey;
+
+            for (uint256 i = 0; i < pathLength; i++) {
+                pathKey = params.path[i];
+                (PoolKey memory poolKey, bool zeroForOne) = pathKey.getPoolAndSwapDirection(currencyIn);
+                amountOut = uint128(_swap(poolKey, zeroForOne, -int256(uint256(amountIn)), 0, pathKey.hookData));
+
+                amountIn = amountOut;
+                currencyIn = pathKey.intermediateCurrency;
+            }
+
+            if (amountOut < params.amountOutMinimum) revert TooLittleReceived();
+        }
+    }
+
+    function _swapExactOutputSingle(IV4Router.ExactOutputSingleParams memory params) private {
+        _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(int128(params.amountOut)),
+            params.sqrtPriceLimitX96,
+            params.hookData
+        );
+    }
+
+    function _swapExactOutput(IV4Router.ExactOutputParams memory params) private {
+        unchecked {
+            // Caching for gas savings
+            uint256 pathLength = params.path.length;
+            uint128 amountIn;
+            uint128 amountOut = params.amountOut;
+            Currency currencyOut = params.currencyOut;
+            PathKey memory pathKey;
+
+            for (uint256 i = pathLength; i > 0; i--) {
+                pathKey = params.path[i - 1];
+                (PoolKey memory poolKey, bool oneForZero) = pathKey.getPoolAndSwapDirection(currencyOut);
+                amountIn = uint128(-_swap(poolKey, !oneForZero, int256(uint256(amountOut)), 0, pathKey.hookData));
```

Comment:
I feel like I commented at some point about documenting/adding a little comment about the return signs but can't find it anywhere.. Anyways I do think it would be nice explaining that the output is guaranteed to be negative for exactOutput in the reciprocal token.

Reply:
have actually added a safecast on this - i think with custom accounting hooks on low liquidity pools its technically not guaranteed